cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -268,78 +282,370 @@ func copyFromGlobHelper(src, dst, globEx string) error {
+	// volume, or a decrypted mapper volume if ubuntu-data is really encrypted
+	// on the physical disk
+	dataDevice string
+	// the device for ubuntu-save that is to be mounted
+	saveDevice string
+
+	// XXX: would be nice to get rid of tracking this here
+	isEncryptedDev bool
+
+	// state for tracking what happens
+	degradedState recoverDegradedState
+}
+
+// TODO: should this be a method on *context ?
+func verifyMountPointCtx(ctx *context, dir, name string) error {
+	matches, err := ctx.disk.MountPointIsFromDisk(dir, ctx.diskOpts)
+	if err != nil {
 		return err
 	}
+	if !matches {
+		return fmt.Errorf("cannot validate boot: %s mountpoint is expected to be from disk %s but is not", name, ctx.disk.Dev())
+	}
+	return nil
+}
 
-	// get the disk that we mounted the ubuntu-seed partition from as a
-	// reference point for future mounts
-	disk, err := disks.DiskFromMountPoint(boot.InitramfsUbuntuSeedDir, nil)
+type stateFunc func(ctx *context) (stateFunc, error)
+
+var (
+	// errStateDone is the error returned when the state machine is done
+	// executing without a critical error
+	errStateDone = fmt.Errorf("state machine done")
+)
+
+type stateMachine struct {
+	current stateFunc
+}
+
+func newStateMachine() *stateMachine {
+	m := &stateMachine{}
+	m.current = m.unlockDataRunKey
+	return m
+}
+
+func (m *stateMachine) execute(ctx *context) error {
+	next, err := m.current(ctx)
+	m.current = next
+	return err
+}
+
+// stateUnlockDataRunKey will try to unlock ubuntu-data with the normal run-mode
+// key, and if it fails, progresses to the next state, which is either:
+// - failed to unlock data, but we know it's an encrypted device -> try to unlock with fallback key
+// - failed to find data at all -> try to unlock save
+// - unlocked data with run key -> mount data
+func (m *stateMachine) unlockDataRunKey(ctx *context) (stateFunc, error) {
+	runModeKey := filepath.Join(boot.InitramfsEncryptionKeyDir, "ubuntu-data.sealed-key")
+	dataDevice, isEncryptedDev, err := secbootUnlockVolumeUsingSealedKeyIfEncrypted(ctx.disk, "ubuntu-data", runModeKey, false)
+	if isEncryptedDev {
+		ctx.diskOpts = &disks.Options{
+			IsDecryptedDevice: true,
+		}
+		ctx.isEncryptedDev = true
+	}
 	if err != nil {
-		return err
+		// we couldn't unlock ubuntu-data with the primary key, or we didn't
+		// find it in the unencrypted case
+		if isEncryptedDev {
+			// we know the device is encrypted, so the next state is to try
+			// unlocking with the fallback key
+			return m.unlockDataFallbackKey, nil
+		}
+
+		// not an encrypted device, so nothing to fall back to try and unlock
+		// data, so just mark it as not found and continue on to try and mount
+		// an unencrypted ubuntu-save directly
+		// TODO:UC20: should we save the specific error in degradedState
+		//            somewhere in addition to logging it?
+		logger.Noticef("failed to find ubuntu-data partition for mounting host data: %v", err)
```

Comment:
I think it'd be useful to have like something like a journal of steps that failed with the respective error messages.

Reply:
Added `ErrorLog []string`